### PR TITLE
Append client version info to graffiti

### PR DESCRIPTION
**Pull request by chong-he**
## Issue Addressed

* #7201 

## Additional Info

Thank you @ethDreamer and @michaelsproul for the help and guidance

### CODE REVIEW

File: beacon_node/http_api/src/produce_block.rs [hunk 1]
```diff
@@ -61,11 +63,21 @@ pub async fn produce_block_v3<T: BeaconChainTypes>(
...
+    let final_graffiti = match query.graffiti_policy {
+        GraffitiPolicy::PreserveUserGraffiti => query.graffiti,
+        GraffitiPolicy::AppendClientVersions => Some(
+            chain
+                .calculate_graffiti()
+                .get_graffiti(query.graffiti)
+                .await,
+        ),
+    };
+
```
Comment (ethDreamer) [inline beacon_node/http_api/src/produce_block.rs:73]: This is the wrong place to do this. This function will call `produce_block_with_verification()` which will call `produce_block_on_state()` which will call the `graffiti_calculator.get_graffiti()` again. I would suggest you pass this argument down the function calls along with the graffiti. So maybe (inside graffiti_calculator.rs) create some new enum:
```rust
enum GraffitiSettings {
    Unspecified,
    Specified {
        graffiti: Graffiti,
        policy: GraffitiPolicy,
    }
}

impl GraffitiSettings {
    fn new(validator_graffiti: Option<Graffiti>, policy: GraffitiPolicy) {
        validator_graffiti.map(|graffiti| Self::Specified {
            graffiti,
            policy,
        })
        .unwrap_or(Self::Unspecified)
    }
}
```
and pass that down through the functions. Then you'd want to change `get_graffiti` to:
```rust
pub async fn get_graffiti(&self, graffiti_settings: GraffitiSettings) -> Graffiti
```

Comment (chong-he) [reply beacon_node/http_api/src/produce_block.rs:73]: Thanks so much for the comment, made some amendment based on your comment. If you have some time to spare, will appreciate for another look to see if this is on the right track @ethDreamer

**Review by jimmygchen — COMMENTED** [review]
🤖 **Automated Review Comments**

## Comment 1: Query Parameter Handling

**File:** beacon_node/http_api/tests/tests.rs:3534

Tests pass `None` as the fifth parameter to `get_validator_blocks_v3`. Please verify that `graffiti_policy: Option<GraffitiPolicy>` was added to `ValidatorBlocksQuery` and is properly deserialized from query parameters.

---

## Comment 2: Complex Graffiti Logic Documentation

**File:** beacon_node/execution_layer/src/engine_api.rs:751

The graffiti truncation logic is complex. Consider adding a comment explaining the design:

```rust
// Graffiti truncation strategy (32 byte limit):
// - User <20 chars: Full version info (12 chars) + space + user
// - User 20-23 chars: 1-byte commits (8 chars) + space + user
// - User 24-27 chars: Client codes only (4 chars) + space + user
// - User 28-29 chars: EL code only (2 chars) + space + user  
// - User ≥30 chars: Preserve user graffiti unchanged
```

---

## Comment 3: Edge Case Test Coverage

**File:** beacon_node/beacon_chain/src/graffiti_calculator.rs:96

Missing test coverage for the edge case where user specifies `--graffiti-policy append-client-versions` but provides empty/no graffiti. Should this append full client version info?

File: beacon_node/http_api/tests/tests.rs [hunk 2]
```diff
@@ -3531,7 +3531,7 @@ impl ApiTester {
 
             let (response, metadata) = self
                 .client
-                .get_validator_blocks_v3_ssz::<E>(slot, &randao_reveal, None, None)
+                .get_validator_blocks_v3_ssz::<E>(slot, &randao_reveal, None, None, None)
```
Comment (jimmygchen) [inline beacon_node/http_api/tests/tests.rs:3684]: 🤖 Please verify that `graffiti_policy: Option<GraffitiPolicy>` was added to `ValidatorBlocksQuery` and is properly deserialized from query parameters.

File: beacon_node/execution_layer/src/engine_api.rs [hunk 3]
```diff
@@ -744,6 +748,49 @@ impl ClientVersionV1 {
                 .unwrap_or("0000")
                 .to_lowercase(),
         );
+
```
Comment (jimmygchen) [inline beacon_node/execution_layer/src/engine_api.rs:785]: 🤖 The graffiti truncation logic here is complex. Consider adding a comment explaining the design:

```rust
// Graffiti truncation strategy (32 byte limit):
// - User <20 chars: Full version info (12 chars) + space + user
// - User 20-23 chars: 1-byte commits (8 chars) + space + user
// - User 24-27 chars: Client codes only (4 chars) + space + user
// - User 28-29 chars: EL code only (2 chars) + space + user  
// - User ≥30 chars: Preserve user graffiti unchanged
```

File: beacon_node/beacon_chain/src/graffiti_calculator.rs [hunk 4]
```diff
@@ -73,11 +93,19 @@ impl<T: BeaconChainTypes> GraffitiCalculator<T> {
     /// 2. Graffiti specified by the user via beacon node CLI options.
     /// 3. The EL & CL client version string, applicable when the EL supports version specification.
     /// 4. The default lighthouse version string, used if the EL lacks version specification support.
-    pub async fn get_graffiti(&self, validator_graffiti: Option<Graffiti>) -> Graffiti {
-        if let Some(graffiti) = validator_graffiti {
-            return graffiti;
+    pub async fn get_graffiti(&self, graffiti_settings: GraffitiSettings) -> Graffiti {
```
Comment (jimmygchen) [inline beacon_node/beacon_chain/src/graffiti_calculator.rs:96]: 🤖 Missing test coverage for the edge case where user specifies `--graffiti-policy append-client-versions` but provides empty/no graffiti. Should this append full client version info?

File: beacon_node/http_api/tests/tests.rs (hunk 2 above)
Comment (chong-he) [reply beacon_node/http_api/tests/tests.rs:3684]: yes, this field is added to the struct:
https://github.com/sigp/lighthouse/pull/7558/files#diff-d43db73dcaf47e9e087bdc0d1f9c6d8db76ac6843d47d1e555b386493b550a52R789

File: beacon_node/beacon_chain/src/graffiti_calculator.rs (hunk 4 above)
Comment (chong-he) [reply beacon_node/beacon_chain/src/graffiti_calculator.rs:96]: The test `check_append_el_version_graffiti_various_length` covers grattifi strings of various length, including empty/no graffiti

File: beacon_node/execution_layer/src/engine_api.rs (hunk 3 above)
Comment (chong-he) [reply beacon_node/execution_layer/src/engine_api.rs:785]: comment added in [c3c163c](https://github.com/sigp/lighthouse/pull/7558/commits/c3c163cf5fb0788db5a75396909f1bacb4c981cd)

File: beacon_node/beacon_chain/src/graffiti_calculator.rs [hunk 5]
```diff
@@ -352,12 +391,110 @@ mod tests {
...
         );
     }
+
+    #[tokio::test]
+    async fn check_append_el_version_graffiti_various_length() {
+        let spec = Arc::new(test_spec::<MinimalEthSpec>());
+        let harness = get_harness(VALIDATOR_COUNT, spec, None);
+
+        let graffiti_vec = vec![
+            // less than 20 characters, example blow is 19 characters
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/graffiti_calculator.rs]:
```suggestion
            // less than 20 characters, example below is 19 characters
```

File: book/src/help_vc.md [hunk 6]
```diff
@@ -221,6 +221,9 @@ Flags:
           automatically enabled for <= 64 validators. Enabling this metric for
           higher validator counts will lead to higher volume of prometheus
           metrics being collected.
+      --graffiti-append
+          When used, client version info will be automatically appended to user
+          custom graffiti.
```
Comment (michaelsproul) [inline book/src/help_vc.md]: We should probably have a note that this should only be used with a Lighthouse BN (due to using a non-standard query parameter).

**Review by michaelsproul — APPROVED** [review]
Approved modulo some very small doc/CLI flag tweaks

File: book/src/help_vc.md (hunk 6 above)
Comment (chong-he) [reply book/src/help_vc.md]: Revised in [63746d8](https://github.com/sigp/lighthouse/pull/7558/commits/63746d8c3ad90eb21c5efbb2885ad078da3b3980)

Also revise the word from append to prepend, as the client version info is attached before user graffiti, not after

**Comment by chong-he** [comment]
Did some testing with Kurtosis local testnet, by adding the following to `network_params.yaml`:
```
vc_extra_params:
      - --graffiti-append
      - --graffiti=testing123
```

Before, graffiti is: `testing123`
After, graffiti is: `GE6978LH49e1 testing123`

File: book/src/help_vc.md (hunk 6 above)
Comment (michaelsproul) [reply book/src/help_vc.md]: Good catch, makes sense!
